Add PDE

nicnys-8 opened the pull request:
Add a Plant Description Engine core system. It works with the latest versions of the Orchestrator and Service registry. It uses version 0.6 of ar:kalix, which is not yet publicly available.

rbocsi commented on plantdescriptionengine/example-plant-description.json:
```diff
@@ -0,0 +1,87 @@
+            ]
+        },
+        {
+            "systemName": "test_system",
```
If this systemName appears as system name in ServiceRegistry, please note that from version 4.4.0 underscore in system names is not valid, Service Registry will reject requests with such system names (see issue #202).

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/pom.xml:
```diff
@@ -0,0 +1,170 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
This pom does not follow the structure of the project, where there is a main pom in the root of the repository and there is a pom for every module that uses the main one as parent. (See any core system for example.) This way a lot a common dependencies can remove from this pom, only the special ones (like kalix) should remain.

The version is confusing. The PDE will release as part of Arrowhead 4.4.0, so its version should be that.
PDE currently does not provide a debian installer (by using jdeb maven plugin).

nicnys-8 replied:
I have updated to POM to match the core system structure, but I have not yet added a debian installer.

nicnys-8 replied:
I have added a debian installer for the PDE. However, I've not yet managed to install any of the new versions of the core systems using the deb installers.

nicnys-8 replied:
I've tested the debian installer on my machine, and it works.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/MonitorInfo.java:
```diff
@@ -0,0 +1,174 @@
+            result += ",metadata=" + Metadata.toString(metadata);
+        }
+
+        return result + "}";
```
Don't find the opening brace, maybe missing.

nicnys-8 replied:
Fixed!

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/MonitorInfo.java:
```diff
@@ -0,0 +1,174 @@
+        return result;
+    }
+
+    public static class Bundle {
```
why public if its only constructor is not public?

nicnys-8 replied:
Instances of the Bundle class were used outside of the class MonitorInfo, but they were only created inside MonitorInfo. However, I've refactored the code now and put the inner class in its own file.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+     * @return HTTP client useful for consuming Arrowhead services.
+     */
+    static HttpClient createHttpClient(final Properties appProps) {
+
```
null check on appProps before using

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+     */
+    static HttpClient createHttpClient(final Properties appProps) {
+
+        final boolean secureMode = Boolean.parseBoolean(getProp(appProps, "server.ssl.enabled"));
```
Generally, it is a good practice to move string literals to constants. The only exceptions are the descriptions, log and error messages.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+        }
+
+        final OwnedIdentity identity = loadIdentity(
+            getProp(appProps, "server.ssl.pde.key-store"),
```
In every other core systems, these keys are the same: server.ssl.key-store, server.ssl.key-store-password, etc. Why include the pde here? It's just confuse the admin of the system.

nicnys-8 replied:
Whoops, that's a remnant from when the PDE used two identities (a separate one for communicating with core systems). Fixed now.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+     * @return An Arrowhead Framework System.
+     */
+    static ArSystem createArSystem(final Properties appProps, final InetSocketAddress serviceRegistryAddress) {
+
```
input check

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+            try (InputStream in = ClassLoader.getSystemResourceAsStream(path)) {
+                keyStore.load(in, password);
+            } catch (NoSuchAlgorithmException | CertificateException | IOException e) {
+                logger.error("Failed to load keystore from resources", e);
```
typo: resources => resource

nicnys-8 replied:
Clarified (changed to _resources directory_)

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+            try {
+                appProps.load(ClassLoader.getSystemResourceAsStream(APP_PROPS_FILENAME));
+            } catch (IOException e) {
+                logger.error("Failed reading " + APP_PROPS_FILENAME + " from system resources.", e);
```
typo: resources => resource

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/PdeMain.java:
```diff
@@ -0,0 +1,381 @@
+            trustStore = TrustStore.from(keyStore);
+        } catch (KeyStoreException e) {
+            logger.error("Failed to read trust store", e);
+            System.exit(1);
```
Everywhere else you use 74 as exit code. I'm not sure this change is intentional.

nicnys-8 replied:
Fixed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/SystemMismatchDetector.java:
```diff
@@ -0,0 +1,274 @@
+
+        if (entrySystem.systemName().isPresent()) {
+            final String entryName = entrySystem.systemName().get();
+            return entryName.equals(registeredSystem.systemName());
```
Please note that the Service Registry normalizes (systemName.trim().toLowerCase()) the system names before storing in the database. So, maybe here you should trim the names too, and use equalsIgnoreCase instead of equals.

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/SystemMismatchDetector.java:
```diff
@@ -0,0 +1,274 @@
+
+        final String alarmSystemName = alarm.getSystemName();
+        final Map<String, String> alarmMetadata = alarm.getMetadata();
+        final boolean namesMatch = alarmSystemName != null && alarmSystemName.equals(system.systemName());
```
My previous comments may be relevant here too.

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/Alarm.java:
```diff
@@ -0,0 +1,134 @@
+        final Map<String, String> nonNullMetadata = metadata == null ? Collections.emptyMap() : metadata;
+        return cause == this.cause &&
+            Objects.equals(systemId, this.systemId) &&
+            Objects.equals(systemName, this.systemName) &&
```
I don't see where the systemName parameter comes from, but if it comes from the Service Registry, then the previously mentioned normalization can cause issues here.

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/AlarmManager.java:
```diff
@@ -0,0 +1,263 @@
+    public List<PdeAlarmDto> getAlarms() {
+        final List<PdeAlarmDto> result = new ArrayList<>();
+
+        final List<Alarm> allAlarms = new ArrayList<>();
```
Maybe add activeAlarms.size() + clearedAlarms.size() as constructor parameters. This way you can avoid the ArrayList internal resizing.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/Alarm.java:
```diff
@@ -0,0 +1,134 @@
+     */
+    public void setAcknowledged(final boolean acknowledged) {
+        this.acknowledged = acknowledged;
+        acknowledgedAt = Instant.now();
```
what if the parameter is false?

nicnys-8 replied:
You're right, I'm removing the possibility of unacknowledging an acknowledged alarm.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/AlarmManager.java:
```diff
@@ -0,0 +1,263 @@
+        final Map<String, String> metadata,
+        final AlarmCause cause
+    ) {
+        synchronized (this) {
```
You should avoid synchronized on this. Instead, you can use a member for this purpose:
`private final Object lock = new Object();`
See details in https://stackoverflow.com/questions/3369287/what-is-the-difference-between-synchronized-on-lockobject-and-using-this-as-the

nicnys-8 replied:
Good point, fixed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/AlarmManager.java:
```diff
@@ -0,0 +1,263 @@
+            .filter(alarm ->
+                alarm.getCause() == AlarmCause.NOT_MONITORABLE &&
+                    Objects.equals(systemId, alarm.getSystemId()) &&
+                    Objects.equals(systemName, alarm.getSystemName()) &&
```
maybe system name normalization can cause issues

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/AlarmManager.java:
```diff
@@ -0,0 +1,263 @@
+        final List<Alarm> cleared = activeAlarms.stream()
+            .filter(alarm ->
+                alarm.getCause() == AlarmCause.NO_PING_RESPONSE &&
+                    systemName.equals(alarm.getSystemName())
```
maybe system name normalization can cause issues

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/consumedservices/orchestrator/dto/ProviderSystem.java:
```diff
@@ -0,0 +1,21 @@
+@DtoReadableAs(JSON)
+@DtoWritableAs(JSON)
+@DtoToString
+public interface ProviderSystem {
```
I'm not sure if you need this here, but in Service Registry systems have authenticationInfo and metadata fields as well.

nicnys-8 replied:
The ProviderSystem class has been removed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/consumedservices/orchestrator/RuleCreator.java:
```diff
@@ -0,0 +1,76 @@
+            .serviceMetadata(producerPort.metadata())
+            .serviceInterfaceName(producerPort.serviceInterface().orElse(DEFAULT_SERVICE_INTERFACE))
+            .serviceDefinitionName(producerPort.serviceDefinition());
+
```
maybe priority?

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/consumedservices/orchestrator/rulebackingstore/SqlRuleStore.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Override
+    public void writeRules(final int plantDescriptionId, final Set<Integer> rules) throws RuleStoreException {
```
No transaction handling. If, for example, the 5th record writing is failed, the first 4 remains in the db. You should turn off automatic commit before the loop, commit after the loop, rollback in the catch block and turn on automatic commit in a finally block.

See https://docs.oracle.com/javase/tutorial/jdbc/basics/transactions.html for example.

nicnys-8 replied:
Fixed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/consumedservices/serviceregistry/SystemTracker.java:
```diff
@@ -0,0 +1,214 @@
+
+        // TODO: Throw an Exception if more than one match is found?
+        for (final SrSystem system : systems) {
+            if (systemName.equals(system.systemName()) && Metadata.isSubset(metadata, system.metadata())) {
```
The system name normalization can cause issues here.

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/pdtracker/PlantDescriptionTracker.java:
```diff
@@ -0,0 +1,336 @@
+            maxId = Math.max(maxId, entry.id());
+            entries.put(entry.id(), entry);
+        }
+        nextId.set(Math.round(maxId + 1));
```
why do you need round here? Every number is int.

nicnys-8 replied:
ID:s used to be floats. Fixed now.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/pdtracker/PlantDescriptionTracker.java:
```diff
@@ -0,0 +1,336 @@
+        final PlantDescriptionEntry oldEntry;
+        final PlantDescriptionEntryDto deactivatedEntry;
+
+        synchronized (this) {
```
see a previous comment about synchronized

nicnys-8 replied:
Fixed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/pdtracker/backingstore/SqlPdStore.java:
```diff
@@ -0,0 +1,161 @@
+    // TODO: We store Plant Descriptions in their raw JSON form.
+    // In the future, we'll want to create separate tables for each subfield
+    // of a Plant Description.
+    private final String SQL_CREATE_TABLE = "CREATE TABLE IF NOT EXISTS plant_description (id INT, plant_description TEXT);";
```
is 64kb enough to store the JSON? If not, set this to MEDIUMTEXT which can store texts up to 64MB.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/pdtracker/backingstore/SqlPdStore.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    @Override
+    public void write(final PlantDescriptionEntryDto entry) throws PdStoreException {
```
Transaction handling

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/PlantDescriptionValidator.java:
```diff
@@ -0,0 +1,446 @@
+    }
+
+    /**
+     * Report an error if the given system's name is illegal.
```
The method checks the systemId, not the name.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/PlantDescriptionValidator.java:
```diff
@@ -0,0 +1,446 @@
+    }
+
+    private void validate(final List<PdeSystem> systems) {
+        for (final PdeSystem system : systems) {
```
Somewhere here maybe you should check that the systemName is valid for SR too. See CommonNamePartVerifier class in arrowhead-common module how we do this.

nicnys-8 replied:
Fixed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/PlantDescriptionValidator.java:
```diff
@@ -0,0 +1,446 @@
+                if (!allEntries.containsKey(includedId)) {
+                    throw new ValidationException("Error in include list: Entry '" + includedId + "' is required by entry '" + nextEntry.id() + "'.");
+                }
+                final var includedEntry = allEntries.get(includedId);
```
My opinion is that using var worsen code readability

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/routehandlers/GetAllPlantDescriptions.java:
```diff
@@ -0,0 +1,161 @@
+
+        PlantDescriptionEntryListDto result = new PlantDescriptionEntryListDto.Builder()
+            .data(entries)
+            .count(entries.size())
```
This count calculation works differently than in the similar endpoints of the other core systems. For example, in Service Registry's /mgmt/systems endpoint the count doesn't contain the size of the data, but the total number of records AFTER filtering but BEFORE paging.

You should move active filtering before the paging, then store the size of the entries in a local variable. And use this variable when you build the result.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/routehandlers/ReplacePlantDescription.java:
```diff
@@ -0,0 +1,91 @@
+                }
+
+                return response
+                    .status(HttpStatus.CREATED)
```
a PUT request shoudn't return with 201, but 200 (or 204).

See https://stackoverflow.com/questions/2342579/http-status-code-for-update-and-delete

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/PingTask.java:
```diff
@@ -0,0 +1,144 @@
+
+public class PingTask extends TimerTask {
+
+    private static final String MONITORABLE_SERVICE_NAME = "monitorable";
```
This constants appears in PdeMonitorService class as well. Maybe you can create a public constant for this somewhere and use that in every place.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/PingTask.java:
```diff
@@ -0,0 +1,144 @@
+
+    private boolean isProvidedBy(SystemRecord provider, PdeSystem system) {
+        boolean metadataMatch = Metadata.isSubset(system.metadata(), provider.metadata());
+        boolean nameMatch = system.systemName().isEmpty() || system.systemName().get().equals(provider.name());
```
possible issues because of system name normalization (if any name comes from the SR)

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/dto/PdeAlarm.java:
```diff
@@ -0,0 +1,147 @@
+        return cleared1.get().compareTo(cleared2.get());
+    };
+
+    String NOT_CLEARED = "not_cleared";
```
maybe this should be a constant

nicnys-8 replied:
A final modifier would be redundant, since all fields in an interface are implicitly public, static and final.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/dto/PdeAlarm.java:
```diff
@@ -0,0 +1,147 @@
+        Objects.requireNonNull(alarms, "Expected alarms.");
+        Objects.requireNonNull(systemName, "Expected systemName.");
+
+        alarms.removeIf(alarm -> alarm.systemName().isEmpty() || !alarm.systemName().get().equals(systemName));
```
possible issues because of system name normalization (if any name comes from the SR)

nicnys-8 replied:
System names are now validated at the time they enter the PDE. This means that the system names used here are trimmed and on the correct format.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/DtoUtils.java:
```diff
@@ -0,0 +1,175 @@
+     * @return A list of connections adhering to the monitor package format.
+     */
+    private static List<ConnectionDto> mgmtToMonitor(final List<? extends Connection> connections) {
+        final List<ConnectionDto> result = new ArrayList<>();
```
initial size should be connections.size()

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/DtoUtils.java:
```diff
@@ -0,0 +1,175 @@
+
+                for (final MonitorInfo.Bundle info : systemInfoList) {
+
+                    final boolean matchesServiceDefinition = info.serviceDefinition.equals(port.serviceDefinition());
```
Similar to system name, serviceDefinitions is also normalized (toLowerCase().trim()) in the Service Registry. So, if any serviceDefinition here comes from the SR, that can cause issue in the comparison.

nicnys-8 replied:
Service definitions are now validated upon entering the PDE.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/GetAllPdeAlarms.java:
```diff
@@ -0,0 +1,193 @@
+            }
+        }
+
+        final Optional<Integer> page = parser.getValue(pageParam);
```
filtering should always precedes paging.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/GetAllPdeAlarms.java:
```diff
@@ -0,0 +1,193 @@
+
+        PdeAlarmListDto result = new PdeAlarmListDto.Builder()
+            .data(alarms)
+            .count(alarms.size())
```
count should be the size of alarms before paging

nicnys-8 replied:
Fixed.

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/GetAllPlantDescriptions.java:
```diff
@@ -0,0 +1,173 @@
+
+    // Filter fields
+    private static final String ACTIVE = "active";
+    private static final String ITEM_PER_PAGE = "item_per_page";
```
A lot of constants here (ITEM_PER_PAGE, PAGE, SORT_FIELD, etc.) appears in multiple classes. You should move these constants (and making them public) into one dedicated class and use everywhere instead of copying into every class.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/GetAllPlantDescriptions.java:
```diff
@@ -0,0 +1,173 @@
+        final Optional<String> sortField = parser.getValue(sortFieldParam);
+        if (sortField.isPresent()) {
+            final String sortDirection = parser.getRequiredValue(directionParam);
+            final boolean sortAscending = "ASC".equals(sortDirection);
```
ASC should be constant as well

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/GetAllPlantDescriptions.java:
```diff
@@ -0,0 +1,173 @@
+            entries = entries.subList(from, to);
+        }
+
+        final Optional<Boolean> active = parser.getValue(activeParam);
```
filtering before paging

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitor/routehandlers/GetAllPlantDescriptions.java:
```diff
@@ -0,0 +1,173 @@
+
+        PlantDescriptionEntryListDto result = new PlantDescriptionEntryListDto.Builder()
+            .data(extendedEntries)
+            .count(extendedEntries.size())
```
count should be the total number of entries before paging

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_monitorable/routehandlers/GetSystemData.java:
```diff
@@ -0,0 +1,44 @@
+        Objects.requireNonNull(request, "Expected request.");
+        Objects.requireNonNull(response, "Expected response.");
+
+        JsonObject data = new JsonObject(new JsonPair("name", new JsonString(systemName)));
```
"name" should be a constant

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/requestvalidation/IntParameter.java:
```diff
@@ -0,0 +1,104 @@
+
+        if (intValue < minValue) {
+            parser.report(new ParseError(
+                "Query parameter '" + name + "' must be greater than " + minValue + ", got " + intValue + "."));
```
not true, intValue==minValue is good too.

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/utils/Metadata.java:
```diff
@@ -0,0 +1,59 @@
+
+        for (final Map.Entry<String, String> entry : sorted.entrySet()) {
+            if (stringBuilder.length() > 0) {
+                stringBuilder.append(",");
```
maybe a space after comma

nicnys-8 replied:
Fixed

rbocsi commented on plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/utils/Metadata.java:
```diff
@@ -0,0 +1,59 @@
+            stringBuilder.append("=");
+            stringBuilder.append(entry.getValue());
+        }
+        return stringBuilder.toString();
```
maybe cut the trailing ", ", if necessary

nicnys-8 replied:
Fixed

rbocsi reviewed (changes requested):
See inline comments

rbocsi commented on documentation/plant-description-engine/plant-description-engine-sosd.md:
```diff
@@ -0,0 +1,488 @@
+	"systems": [
+		{
+			"systemName": "Service Registry",
+			"systemId": "service_registry",
```
Please note that if the systemId in PDE has any connection with system's name in Service Registry, then using underscores can be a problem. Since Arrowhead 4.4.0 system's name has to match a particular regexp (see issue #202 or CommonNamePartVerifier class). Because of this we have to rename service_registry to serviceregistry.

nicnys-8 replied:
Fixed, together with a bunch of other invalid system names in the docs.

rbocsi commented on plantdescriptionengine/pom.xml:
```diff
@@ -0,0 +1,168 @@
+            <artifactId>netty-all</artifactId>
+            <version>4.1.52.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>1.4.200</version>
+        </dependency>
```
if you using this only in tests, then the dependency is already in the parent pom.

nicnys-8 replied:
Removed it.

rbocsi commented on plantdescriptionengine/src/deb/control/control:
```diff
@@ -0,0 +1,10 @@
+Section: contrib/java
+Priority: optional
+Architecture: all
+Maintainer: Svetlin Tanyi <szvetlin@aitia.ai>
```
Maintainer should be somebody on your side.

nicnys-8 replied:
Changed to my name and email.

rbocsi commented on plantdescriptionengine/src/deb/control/postrm:
```diff
@@ -0,0 +1,57 @@
+      ${SYSTEM_DIR}/log4j2.xml \
+      ${SYSTEM_DIR}/application.properties \
+      ${SYSTEM_DIR}/${SYSTEM_NAME}.p12 \
+      ${AH_SYSTEMS_DIR}/sysop/sysop.p12
```
deleting sysop is not PDE's responsibility, so delete this line with the \ in the previous line

nicnys-8 replied:
Fixed.

rbocsi commented on plantdescriptionengine/src/deb/control/postrm:
```diff
@@ -0,0 +1,57 @@
+      ${AH_SYSTEMS_DIR}/sysop/sysop.p12
+
+    rmdir ${SYSTEM_DIR} 2>/dev/null || true
+    rmdir ${AH_SYSTEMS_DIR}/sysop 2>/dev/null || true
```
deleting sysop is not PDE's responsibility, so delete this line too

nicnys-8 replied:
Fixed.

rbocsi commented on scripts/plant_description_engine_privileges.sql:
```diff
@@ -0,0 +1,13 @@
+
+GRANT ALL PRIVILEGES ON `arrowhead`.`pde_rule` TO 'plant_description_engine'@'localhost';
+GRANT ALL PRIVILEGES ON `arrowhead`.`plant_description` TO 'plant_description_engine'@'localhost';
+
```
add privilege to log table to as your log4j config wants it.
GRANT ALL PRIVILEGES ON \`arrowhead\`.\`logs\` TO 'plant_description_engine'@'localhost';`

nicnys-8 replied:
Fixed.

rbocsi commented on scripts/plant_description_engine_privileges.sql:
```diff
@@ -0,0 +1,13 @@
+
+GRANT ALL PRIVILEGES ON `arrowhead`.`pde_rule` TO 'plant_description_engine'@'%';
+GRANT ALL PRIVILEGES ON `arrowhead`.`plant_description` TO 'plant_description_engine'@'%';
+
```
add privilege to log table to as your log4j config wants it.
GRANT ALL PRIVILEGES ON \`arrowhead\`.\`logs\` TO 'plant_description_engine'@'%';

nicnys-8 replied:
Fixed.

rbocsi reviewed (changes requested):
I found some further issues (see comments). I also run the tests and two of them failed on my environment, see log:
[mvn.log](https://github.com/arrowhead-f/core-java-spring/files/6638383/mvn.log)

nicnys-8 commented:
I've resolved the latest issues.

The test failures were timing related. I fixed one of them, and removed a couple of others that were a bit trickier to fix. I'll think of a way to rewrite them as well, but I'll just leave them out for now.
